src/core/jsonschema/jsonschema_vocabularies.cc

```diff
@@ -0,0 +1,239 @@
+  if (uri == "http://json-schema.org/draft-07/hyper-schema#")
+    return KnownVocabulary::Draft07Hyper;
+
+  // 2019-09 vocabularies
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/core")
+    return KnownVocabulary::Draft201909Core;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/applicator")
+    return KnownVocabulary::Draft201909Applicator;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/validation")
+    return KnownVocabulary::Draft201909Validation;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/meta-data")
+    return KnownVocabulary::Draft201909MetaData;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/format")
+    return KnownVocabulary::Draft201909Format;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/content")
+    return KnownVocabulary::Draft201909Content;
+  if (uri == "https://json-schema.org/draft/2019-09/vocab/hyper-schema")
+    return KnownVocabulary::Draft201909HyperSchema;
+
+  // 2020-12 vocabularies
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/core")
+    return KnownVocabulary::Draft202012Core;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/applicator")
+    return KnownVocabulary::Draft202012Applicator;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/unevaluated")
+    return KnownVocabulary::Draft202012Unevaluated;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/validation")
+    return KnownVocabulary::Draft202012Validation;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/meta-data")
+    return KnownVocabulary::Draft202012MetaData;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/format-annotation")
+    return KnownVocabulary::Draft202012FormatAnnotation;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/format-assertion")
+    return KnownVocabulary::Draft202012FormatAssertion;
+  if (uri == "https://json-schema.org/draft/2020-12/vocab/content")
+    return KnownVocabulary::Draft202012Content;
+
+  return std::nullopt;
+}
+} // anonymous namespace
+
+// Vocabularies struct implementations
+auto sourcemeta::core::Vocabularies::contains(std::string_view uri) const
+    -> bool {
+  const auto maybe_known = uri_to_known_vocabulary(uri);
+  if (maybe_known.has_value()) {
+    const auto flag = static_cast<uint32_t>(maybe_known.value());
+    // Presence is recorded when either enabled or disabled
+    return ((enabled_known | disabled_known) & flag) != 0;
+  }
+  // Fallback to custom vocabularies
+  const auto it = custom.find(std::string{uri});
+  return it != custom.end() && it->second;
+}
+
+auto sourcemeta::core::Vocabularies::insert(
+    const std::pair<JSON::String, bool> &entry) -> void {
+  const auto maybe_known = uri_to_known_vocabulary(entry.first);
+  if (maybe_known.has_value()) {
+    const auto flag = static_cast<uint32_t>(maybe_known.value());
+    if (entry.second) {
+      enabled_known |= flag;
+      disabled_known &= ~flag; // Clear from disabled
```

Comment:
```suggestion
      // Clear from disabled
      disabled_known &= ~flag;
```

Let's put code comments ABOVE the lines for consistency